allennlp/tests/semparse/domain_languages/quarel_language_test.py

```diff
@@ -0,0 +1,127 @@
+from allennlp.semparse import ExecutionError
+
+class QuaRelLanguageTest(AllenNlpTestCase):
+    def setUp(self):
+        super().setUp()
+        self.language = QuaRel()
+
+    def test_constant_logical_form(self):
+        assert self.language.execute('world2').number == 2
+        with pytest.raises(ExecutionError, match='Unrecognized constant'):
+            self.language.execute('"and"')
+
+    def test_infer_quarel(self):
+        assert self.language.execute(('(infer (speed higher world1) (friction higher world1) '
+                                      '(friction lower world1))')) == 1
+
+        assert self.language.execute(('(infer (speed higher world2) (friction higher world1) '
+                                      '(friction lower world1))')) == 0
+
+        # Both answer options are correct.
+        assert self.language.execute(('(infer (speed higher world2) (friction higher world1) '
+                                      '(friction higher world1))')) == -2
+
+        # Neither answer option is correct.
+        assert self.language.execute(('(infer (speed higher world2) (friction higher world2) '
+                                      '(friction higher world2))')) == -1
+
+        assert self.language.logical_form_to_action_sequence(('(infer (speed higher world1) '
+                                                              '(friction higher world1) '
+                                                              '(friction lower world1))')) == \
+                ['@start@ -> int',
+                 'int -> [<QuaRelType,QuaRelType,QuaRelType:int>, QuaRelType, QuaRelType, '
+                 'QuaRelType]',
+                 '<QuaRelType,QuaRelType,QuaRelType:int> -> infer',
+                 'QuaRelType -> [<Direction,World:QuaRelType>, Direction, World]',
+                 '<Direction,World:QuaRelType> -> speed',
+                 'Direction -> higher',
+                 'World -> world1',
+                 'QuaRelType -> [<Direction,World:QuaRelType>, Direction, World]',
+                 '<Direction,World:QuaRelType> -> friction',
+                 'Direction -> higher',
+                 'World -> world1',
+                 'QuaRelType -> [<Direction,World:QuaRelType>, Direction, World]',
+                 '<Direction,World:QuaRelType> -> friction',
+                 'Direction -> lower',
+                 'World -> world1']
+
+    def test_infer_quaval(self):
+        assert self.language.execute(('(infer (and (thickness low world1) '
+                                      '(thickness high world2)) '
+                                      '(strength lower world1) '
+                                      '(strength lower world2))')) == 0
+
+        assert self.language.execute(('(infer (and (thickness low world1) '
+                                      '(thickness high world2)) '
+                                      '(strength lower world2) '
+                                      '(strength lower world1))')) == 1
```

Comment:
It looks like you'll currently crash if you use an `and` and the two relations aren't compatible.  You should have a check for this case.